Dockerfile

```diff
@@ -0,0 +1,138 @@
+# ARG base_image=ubuntu:18.04
+ARG base_image=python:3.8
+FROM $base_image
+
+USER root
+
+# Version strings and default paths
+ENV SPARK_VERSION=2.4.5 \
+    HADOOP_VERSION=2.7 \
+    HADOOP_HOME=/usr/local/hdp \
+    SPARK_HOME=/usr/local/spark \
+    DBT_HOME=/usr/local/dbt-spark \
+    HOME=/home
+WORKDIR $HOME
+
+### Use spark/conf instead of hadoop/shared/conf:
+ENV HADOOP_CONF_DIR /usr/local/spark/conf
+
+# Env tuning
+ENV PYTHONUNBUFFERED=1 \
+    PYTHONIOENCODING=utf-8 \
+    LANG=C.UTF-8
+
+# Install core libraries and dependencies
+RUN apt-get -y update && \
+    apt-get install -y \
+    apt-transport-https \
+    apt-utils \
+    ca-certificates \
+    ca-certificates-java \
+    iptables \
+    libcurl4-openssl-dev \
+    libsasl2-dev \
+    libssl-dev \
+    libxml2-dev \
+    lxc \
+    openssh-client \
+    net-tools \
+    wget && \
+    rm -rf /var/lib/apt/lists/*
+
+# Install Java
+RUN apt-get -y update && \
+    apt-get install -y \
+    default-jre && \
+    rm -rf /var/lib/apt/lists/* && \
+    echo JAVA_HOME=$JAVA_HOME && \
+    echo "which java=`which java`" && \
+    java -version
+
+# RUN update-alternatives --config java
+# RUN apt-get -y update && \
+#     apt-get install -y \
+#     -t jessie-backports \
+#     openjdk-8-jre-headless && \
+#     rm -rf /var/lib/apt/lists/*
+
+# Download spark
+RUN cd /tmp && \
+    SPARK_NAME=spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION} && \
+    curl -o $SPARK_NAME.tgz http://mirrors.sonic.net/apache/spark/spark-${SPARK_VERSION}/$SPARK_NAME.tgz && \
+    tar -xzf $SPARK_NAME.tgz -C /usr/local --owner root --group root --no-same-owner && \
+    rm $SPARK_NAME.tgz && \
+    cd /usr/local && \
+    ln -s $SPARK_NAME spark
+
+# Spark config
+ENV PYTHONPATH=$SPARK_HOME/python:$SPARK_HOME/python/lib/py4j-0.10.7-src.zip \
+    SPARK_OPTS="--driver-java-options=-Xms1024M --driver-java-options=-Xmx4096M --driver-java-options=-Dlog4j.logLevel=info" \
+    PATH=$PATH:$SPARK_HOME/bin
+
+# Install mysql server and drivers (for hive metastore)
```

Comment:
should we separate the mysql bit in a docker-compose ?

Reply:
I would love to remove the mysql dependency, but I don't feel good about the default experience when this is omitted. I had previously leaned on the default 'derby' implementation for the spark metastore but the problem is that derby falls down immediately as soon as you try to run thrift, which is currently a hard requirement for dbt-spark.

Reply:
but right now we have the MySQL instance running directly inside the same container right ? I'm just wondering if we should remove just the server part of it, and use an existing MySQL docker image, then connect both through docker-compose

Reply:
@dmateusp - Yes, we can certainly do that - remove the mysql server part and use a myswl docker image but (rightly or wrongly) I was really hoping to have a single image "just work" without introducing the extra overhead of docker compose and managing multiple container lifecycles. This might be misguided on my part, but it still seems that would be valueable for a number of use cases.

If it is important to remove the mysql dependency (moving it outside the container), what if we take a two-pronged approach here:

1. As proposed, we remove the mysql-server aspect from this base Dockerfile and instead refactor as a docker-compose.yml. 
2. To support any use cases where having a single standalone docker image is important, I can wrap the core image in second downstream Dockerfile which comes with mysql bootstrapped into the image as the built-in metadata store and "just works" out of box when executed directly via `docker run`.

@dmateusp - Would this be an okay approach? We can provide a viable `docker run` experience (using the second image) while also honoring the fact that docker-compose.yml is "the right way" to support RDMS backends. (In the long run, I would love to find a solution that can host thrift locally without requiring mysql.)

Reply:
Interesting! I often found that docker-compose allowed me to hide complexity from users because I could set some environment variables, ports and mount volumes by default

The reason I like having databases in a separate container (even for local dev environments) is that they can be parameterized easily, if you have a look at the mysql image for example https://hub.docker.com/_/mysql, you can see there's a complete doc and it tells where volumes should be mounted to persist the data, and what environment variables are available

I think we might lack an actual example usage to compare set-ups, do you think you could add just 1 integration test, or/and a README example section ?

Reply:
@dmateusp - Absolutely, that sounds like a great next step. I'll focus on adding some type of integration test that calls dbt-spark, and a readme with usage example. Thanks!